clang/lib/StaticAnalyzer/Checkers/SimpleBoundsChecker.cpp

```diff
@@ -0,0 +1,454 @@
+// 2. It uses Z3 for handling complex non-concrete bounds constraint. Clang
+//    should be compiled with the Z3 solver enabled.
+//
+// Debug Preprocessor Flags:
+// - DEBUG_DUMP: dumps memory regions, bounds and index expressions,
+//               generated SMT formulas, and some return on failure causes
+// - DEBUG_VISITORS: dumps visited nodes when traversing expressions
+//                   and SVals in replaceSVal and getSymExpr
+//===----------------------------------------------------------------------===//
+
+#include "clang/StaticAnalyzer/Checkers/BuiltinCheckerRegistration.h"
+#include "clang/StaticAnalyzer/Core/BugReporter/BugType.h"
+#include "clang/StaticAnalyzer/Core/Checker.h"
+#include "clang/StaticAnalyzer/Core/CheckerManager.h"
+#include "clang/StaticAnalyzer/Core/PathSensitive/CheckerContext.h"
+#include "clang/StaticAnalyzer/Core/PathSensitive/ExprEngine.h"
+#include "clang/StaticAnalyzer/Core/PathSensitive/SMTConv.h"
+#include "clang/StaticAnalyzer/Core/PathSensitive/SMTSolver.h"
+
+#include "clang/AST/ASTConsumer.h"
+#include "clang/AST/RecursiveASTVisitor.h"
+#include "clang/StaticAnalyzer/Core/PathSensitive/SValVisitor.h"
+
+#include <string>
+
+#define DEBUG_DUMP 0
+#define DEBUG_VISITORS 0
+
+using namespace clang;
+using namespace ento;
+
+namespace {
+  class SimpleBoundsChecker : public Checker<check::Location> {
+    mutable std::unique_ptr<BuiltinBug> BT;
+
+    // Replaces the symbol 'from' with the symbol 'to' in the symbolic expression 'E'
+    SVal replaceSVal(ProgramStateRef state, SVal E, SVal from, SVal to) const;
+
+    // Generates a symbolic expression out of the given bounds expression.
+    //  The terms (variables) should have symbolic values already
+    const SymExpr* getSymExpr(ProgramStateRef state, const BoundsExpr *bounds, const LocationContext *LCtx, SValBuilder &SVB) const;
+
+    void reportOutofBoundsAccess(ProgramStateRef outBound, const Stmt *LoadS, CheckerContext &C) const;
+
+
+    public:
+    void checkLocation(SVal l, bool isLoad, const Stmt *S, CheckerContext &C) const;
+  };
+}
+
+void SimpleBoundsChecker::checkLocation(SVal l, bool isLoad, const Stmt *LoadS,
+    CheckerContext &C) const {
+
+  const MemRegion *R = l.getAsRegion();
+  if (!R)
+    return;
+
+  const ElementRegion *ER = dyn_cast<ElementRegion>(R);
+  if (!ER)
+    return;
+
+#if DEBUG_DUMP
+  llvm::errs() << "Element's Region:\n";
+  ER->dump();
+  llvm::errs() << "\n";
+  llvm::errs() << "Element's Base Region:\n";
+  ER->getBaseRegion()->dump();
+  llvm::errs() << "\n";
+#endif
+
+  // Getting pointers to manager objects
+  ProgramStateRef state = C.getState();
+  ProgramStateManager &SM = state->getStateManager();
+  SValBuilder &svalBuilder = SM.getSValBuilder();
+  ASTContext &Ctx = svalBuilder.getContext();
+
+  // Get the index of the accessed element.
+  DefinedOrUnknownSVal Idx = ER->getIndex().castAs<DefinedOrUnknownSVal>();
+
+  if (Idx.isZeroConstant())
```

Comment:
Does this mean we skip checking if the Idx value is 0?   That can fail for array_ptrs with empty bounds.